include/envoy/stats/refcount_ptr.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <atomic>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Implements a reference-counted pointer to a class, so that its external usage
+// model is identical to std::shared_ptr, but the reference count itself is held
+// in the class. The class is expected to implement three methods:
+//    void incRefCount()
+//    bool decRefCount()  -- returns true if the reference count goes to zero.
+//    uint32_t use_count()
+// It may implement them by delegating to RefcountHelper (see below), or by
+// inheriting from RefcountInterface (see below).
+template <class T> class RefcountPtr {
+public:
+  RefcountPtr() : ptr_(nullptr) {}
+  RefcountPtr(T* ptr) : ptr_(ptr) {
+    if (ptr_ != nullptr) {
+      ptr_->incRefCount();
+    }
+  }
+
+  RefcountPtr(const RefcountPtr& src) { set(src.get()); }
+
+  // Allows RefcountPtr<BaseClass> foo = RefcountPtr<DerivedClass>.
+  template <class U> RefcountPtr(const RefcountPtr<U>& src) { set(src.get()); }
```

Comment:
Any copy constructor of this type seems inherently not-safe and/or hard to reason about, given that now we have 2 things pointing to the same memory with potentially independent ref-counts? I guess the idea is that each thing should have independent inc/dec semantics so we shouldn't wind up with mismatched counts? If that's the case can you clarify that in comments?

Reply:
actually I kind of feel like this pattern is inherently safer than std::shared_ptr, where the reference-count is kept outside the object. E,g. this is a crasher:
```
  std::shared_ptr<std::string> a = std::make_shared<std::string>("x");
  std::shared_ptr<std::string> b(a.get());
```
With ReffcountPtr this is safe.

Of course I fully understand that it's hard to get confidence in some pointer-lifetime-sensitive code like this, so I'm exploring if I can get this PR to work with shared_ptr, with less memory usage improvement.

Reply:
Sorry I re-read your comment and did not really address it.

This particular construct also exists in std::shared_ptr and is safe there too. The same underlying object is used; the same ref-counts are used.  But I'll add some comments.

This is currently needed in thread_local_store.cc and a few other places for up-casting shared pointers of various flavors of histograms. It's not excessively difficult to clean that up but it makes the PR harder to reason about.